File: solutions/elixir/saddle-points/1/lib/saddle_points.ex
defmodule SaddlePoints do
  @doc """
  Parses a string representation of a matrix
  to a list of rows
  """
  @spec rows(String.t()) :: [[integer]]
  def rows(str) do
    str
    |> String.split("\n", trim: true)
    |> Enum.map(&row/1)
  end

  defp row(str) do
    str
    |> String.split(" ", trim: true)
    |> Enum.map(&String.to_integer/1)
  end

  @doc """
  Parses a string representation of a matrix
  to a list of columns
  """
  @spec columns(String.t()) :: [[integer]]
  def columns(str) do
    str |> rows() |> Enum.zip_with(& &1)
  end

  @doc """
  Calculates all the saddle points from a string
  representation of a matrix
  """
  @spec saddle_points(String.t()) :: [{integer, integer}]
  def saddle_points(str) do
    row_maxs = str |> rows() |> Enum.map(&Enum.max/1) |> Enum.with_index(1)
    col_mins = str |> columns() |> Enum.map(&Enum.min/1) |> Enum.with_index(1)

    for {row_max, row} <- row_maxs,
        {col_min, col} <- col_mins,
        row_max == col_min,
        do: {row, col}
  end
end

File: solutions/ruby/saddle-points/1/saddle_points.rb
class Grid
  def self.saddle_points(matrix)
    return [] if matrix.empty?

    row_maxima = {}
    col_minima = {}

    # Find maximum for each row
    matrix.each_with_index do |row, row_idx|
      row_maxima[row_idx] = row.max
    end

    # Find minima for each col
    (0..matrix[0].length).each do |col_idx|
      values_in_col = []
      matrix.each do |row|
        values_in_col << row[col_idx]
      end

      col_minima[col_idx] = values_in_col.min
    end

    saddle_points = []

    matrix.each_with_index do |row, row_idx|
      row.each_with_index do |value, col_idx|
        if row_maxima[row_idx] == value && col_minima[col_idx] == value
          saddle_points << { 'row' => row_idx + 1, 'column' => col_idx + 1 }
        end
      end
    end

    saddle_points
  end
end
